src/meta_memcache/base/base_cache_pool.py

```diff
@@ -0,0 +1,210 @@
+import base64
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Dict, Optional, Set, Tuple, Union
+
+from meta_memcache.base.base_serializer import BaseSerializer
+from meta_memcache.base.base_write_failure_tracker import BaseWriteFailureTracker
+from meta_memcache.base.connection_pool import ConnectionPool
+from meta_memcache.base.memcache_socket import MemcacheSocket
+from meta_memcache.errors import MemcacheError, MemcacheServerError
+from meta_memcache.protocol import (
+    ENDL,
+    Conflict,
+    Flag,
+    IntFlag,
+    Key,
+    MetaCommand,
+    Miss,
+    NotStored,
+    Success,
+    TokenFlag,
+    Value,
+)
+from meta_memcache.settings import MAX_KEY_SIZE
+
+
+class BaseCachePool(ABC):
+    def __init__(
+        self,
+        serializer: BaseSerializer,
+        binary_key_encoding_fn: Callable[[Key], bytes],
+        write_failure_tracker: Optional[BaseWriteFailureTracker] = None,
+    ) -> None:
+        self._serializer = serializer
+        self._binary_key_encoding_fn = binary_key_encoding_fn
+        self._write_failure_tracker = write_failure_tracker
+
+    @abstractmethod
+    def _get_pool(self, key: Key) -> ConnectionPool:
+        pass
```

Comment:
What do you think in throwing and `NotImplementedException` here?

Reply:
It's abstract, you can't instantiate the class without overriding it. But I can use the ... style that is slightly more common for abstract methods.

Reply:
Oh, I dunno why I thought it was also a class method. nvm